docker/docker_image_dest.go

```diff
@@ -131,15 +132,49 @@ func (d *dockerImageDestination) HasThreadSafePutBlob() bool {
 // to any other readers for download using the supplied digest.
 // If stream.Read() at any time, ESPECIALLY at end of input, returns an error, PutBlob MUST 1) fail, and 2) delete any data stored so far.
 func (d *dockerImageDestination) PutBlob(ctx context.Context, stream io.Reader, inputInfo types.BlobInfo, cache types.BlobInfoCache, isConfig bool) (types.BlobInfo, error) {
+	digester, stream := putblobdigest.DigestIfCanonicalUnknown(stream, inputInfo)
+	sizeCounter := &sizeCounter{}
+	stream = io.TeeReader(stream, sizeCounter)
+
+	// If requested, precompute the blob digest to prevent uploading layers that already exist on the registry.
+	// This functionality is particularly useful when BlobInfoCache has not been populated with compressed digests,
+	// the source blob is uncompressed, and the destination blob is being compressed "on the fly".
+	if inputInfo.Digest == "" && d.c.sys.DockerRegistryPushPrecomputeDigests {
```

Comment:
We already have almost the same code in `docker/internal/tarfile.PutBlob`.

Maybe that could be factored out into a separate helper (`internal/{streamdigest,savestream,…}`, I don’t know what’s a good name but I also don’t worry about internal names too much) — that would make the size of the feature in this file < 5 lines and pretty much make my concerns about taking on long-term maintenance a rarely-used feature go completely away.

Reply:
Sounds good. I'll work on a refactor and will update the PR.